docs/RFCS/20210527_changefeed_http_sink.md

```diff
@@ -0,0 +1,202 @@
+- Status: draft
+- Start Date: 2021-05-27
+- Authors: Ryan Min
+- RFC PR: #65927
+- Cockroach Issue: #65816
+
+# Summary
+
+Currently, creating a changefeed with an HTTP endpoint for a sink 
+(`experimental-http`) uses the cloud storage sink implementation (see 
+[sink_cloudstorage.go](../../pkg/ccl/changefeedccl/sink_cloudstorage.go)). Many 
+Enterprise customers have expressed an interest in a webhook-type HTTP sink as
+opposed to the current implementation, which should be reserved for traditional
+cloud storage destinations (i.e. GCS/S3). 
+
+# Technical design
+
+## High-level Design
+
+The HTTP sink implements the existing 
+[`Sink`](../../pkg/ccl/changefeedccl/sink.go) interface with the same Sink 
+functions `EmitRow()`, `Flush()`, etc. and should offer the same guarantees 
+i.e. ordering, at-least-once delivery as other sinks. However, this sink 
+differs slightly in that it iss webhook-based and therefore stateless, so the 
+endpoints don't have to be health-checked and maintained throughout the 
+lifespan of the changefeed. So `Dial()` and `Close()`in the `Sink` interface 
+are no-ops in this scenario, at least for the sink itself (channel still needs 
+to be closed).
+
+The majority of the http connection logic should be wrapped in an 
+[`httputil.Client`](../../pkg/util/httputil/client.go) object (wrapper around Go's 
+[`http.Client`](https://golang.org/src/net/http/client.go#L57)) which will
+handle most of the TLS and other config setup mentioned below. The HTTP sink
+struct might look something like this:
+
+```
+type httpSink struct {
+  u sinkURL
+  client *httputil.Client
+	mu struct {
+		syncutil.Mutex
+		inflight int64
+		flushCh chan struct{}
+	}
+}
+```
+
+## Concurrency
+
+Each `POST` to the HTTP endpoint is executed asynchronously through calls to 
+`EmitRow()` and `EmitResolvedTimestamp()`. We use a channel here, `flushCh`,
+which we will send the http response object once it is received, or the error
+(if one occurs). We use a channel as opposed to something like a `WaitGroup` to
+receive a response (which may be needed for ack functionality later on) or an 
+error.
+
+```
+func (s *httpSink) EmitRow (...) error {
+  ...
+  res, err := s.client.Post(...)
+  s.mu.Lock()
+  s.mu.inflight++
+  s.mu.Unlock()
+  go func() {
+    s.mu.flushCh <- struct {
+      res *http.Response
+      err error
+    } {
+      res: res,
+      err: err,
+    }
+  }
+  ...
+  return err
+}
+```
+
+On the `Flush()` side, the the `mu` field can act as a semaphore, and allow
+reading from the channel until the sink has been flushed.
+
+```
+func (s *httpSink) Flush (...) error {
+  ...
+  s.mu.Lock()
+  defer s.mu.Unlock()
+  for s.mu.inflight > 0 {
+    select {
+      case ctx.Done():
+        return ctx.Err()
+      case flushed := <- s.mu.flushCh:
+        s.mu.inflight--
+        // can add additional ack logic here with flushed.res if needed
+        log.Infof("Sink returned response: %v", flushed.res)
+  }
```

Comment:
I think you should pair down or downright remove those code examples.  RFS should talk about how/why/general mechanism; high level pseudo code like is okay -- but this is a bit too detailed